### PR TITLE
Add user offline tuning

### PR DESCRIPTION
**Pull request by Jay0521**
The mechanism is designed for [SWDEV-489120](https://ontrack-internal.amd.com/browse/SWDEV-489120).

Hipblaslt-bench (get-all) is used to find the best kernel for a problem, and the winner result will be stored into file, by setting the environment variable HIPBLASLT_TUNING_FILE=path, where path points to the stored file.

If the output is stored in a file, the results can be used to override default kernel selection with the kernels found, by setting the environment variable HIPBLASLT_TUNING_OVERRIDE_FILE
=path, where path points to the stored file.

- Before Tuning

```
./hipBLASLt/build/debug/clients/staging/hipblaslt-bench --api_method cpp -m 64 -n 64 -k 1024 --lda 64 --ldb 1024 --ldc 64 --ldd 64  --stride_a 0 --stride_b 0 --stride_c 0 --stride_d 0  --alpha 1.100000 --beta 0.900000 --transA N --transB N --batch_count 1  --a_type f32_r --b_type f32_r --c_type f32_r --d_type f32_r --scale_type f32_r --bias_type f32_r   --compute_type f32_r --print_kernel_info -i 1000 -j 1000 --rotating 512
```
```
transA,transB,grouped_gemm,batch_count,m,n,k,alpha,lda,stride_a,beta,ldb,stride_b,ldc,stride_c,ldd,stride_d,a_type,b_type,c_type,d_type,compute_type,scaleA,scaleB,scaleC,scaleD,amaxD,activation_type,bias_vector,bias_type,rotating_buffer,hipblaslt-Gflops,hipblaslt-GB/s,us,soulution_index

N,N,0,1,64,64,1024,1.1,64,65536,0.9,1024,65536,64,4096,64,4096,f32_r,f32_r,f32_r,f32_r,f32_r,0,0,0,0,0,none,0,f32_r,512,414.395,25.6286,20.243

--Solution index: 48934
...
```

- Tuning

```
export HIPBLASLT_TUNING_FILE=tuning.txt
```

```
./hipBLASLt/build/debug/clients/staging/hipblaslt-bench --api_method cpp -m 64 -n 64 -k 1024 --lda 64 --ldb 1024 --ldc 64 --ldd 64  --stride_a 0 --stride_b 0 --stride_c 0 --stride_d 0  --alpha 1.100000 --beta 0.900000 --transA N --transB N --batch_count 1  --a_type f32_r --b_type f32_r --c_type f32_r --d_type f32_r --scale_type f32_r --bias_type f32_r   --compute_type f32_r --print_kernel_info
```
```
Winner: transA,transB,grouped_gemm,batch_count,m,n,k,alpha,lda,stride_a,beta,ldb,stride_b,ldc,stride_c,ldd,stride_d,a_type,b_type,c_type,d_type,compute_type,scaleA,scaleB,scaleC,scaleD,amaxD,activation_type,bias_vector,bias_type,rotating_buffer,hipblaslt-Gflops,hipblaslt-GB/s,us,soulution_index

N,N,0,1,64,64,1024,1.1,64,65536,0.9,1024,65536,64,4096,64,4096,f32_r,f32_r,f32_r,f32_r,f32_r,0,0,0,0,0,none,0,f32_r,512,903.229,55.8608,9.28735,49257
```

- Override

```
export HIPBLASLT_TUNING_OVERRIDE_FILE=tuning.txt
```

```
./hipBLASLt/build/debug/clients/staging/hipblaslt-bench --api_method cpp -m 64 -n 64 -k 1024 --lda 64 --ldb 1024 --ldc 64 --ldd 64  --stride_a 0 --stride_b 0 --stride_c 0 --stride_d 0  --alpha 1.100000 --beta 0.900000 --transA N --transB N --batch_count 1  --a_type f32_r --b_type f32_r --c_type f32_r --d_type f32_r --scale_type f32_r --bias_type f32_r   --compute_type f32_r --print_kernel_info  --algo_method heuristic --requested_solution 3
```

```
[0]:transA,transB,grouped_gemm,batch_count,m,n,k,alpha,lda,stride_a,beta,ldb,stride_b,ldc,stride_c,ldd,stride_d,a_type,b_type,c_type,d_type,compute_type,scaleA,scaleB,scaleC,scaleD,amaxD,activation_type,bias_vector,bias_type,rotating_buffer,hipblaslt-Gflops,hipblaslt-GB/s,us
    N,N,0,1,64,64,1024,1.1,64,65536,0.9,1024,65536,64,4096,64,4096,f32_r,f32_r,f32_r,f32_r,f32_r,0,0,0,0,0,none,0,f32_r,512,906.538,56.0654,9.25345
    --Solution index: 49257
....

[1]:transA,transB,grouped_gemm,batch_count,m,n,k,alpha,lda,stride_a,beta,ldb,stride_b,ldc,stride_c,ldd,stride_d,a_type,b_type,c_type,d_type,compute_type,scaleA,scaleB,scaleC,scaleD,amaxD,activation_type,bias_vector,bias_type,rotating_buffer,hipblaslt-Gflops,hipblaslt-GB/s,us
    N,N,0,1,64,64,1024,1.1,64,65536,0.9,1024,65536,64,4096,64,4096,f32_r,f32_r,f32_r,f32_r,f32_r,0,0,0,0,0,none,0,f32_r,512,411.904,25.4745,20.3655
    --Solution index: 48934
...

[2]:transA,transB,grouped_gemm,batch_count,m,n,k,alpha,lda,stride_a,beta,ldb,stride_b,ldc,stride_c,ldd,stride_d,a_type,b_type,c_type,d_type,compute_type,scaleA,scaleB,scaleC,scaleD,amaxD,activation_type,bias_vector,bias_type,rotating_buffer,hipblaslt-Gflops,hipblaslt-GB/s,us
    N,N,0,1,64,64,1024,1.1,64,65536,0.9,1024,65536,64,4096,64,4096,f32_r,f32_r,f32_r,f32_r,f32_r,0,0,0,0,0,none,0,f32_r,512,394.991,24.4285,21.2375
    --Solution index: 49003

...
```

```
Winner: transA,transB,grouped_gemm,batch_count,m,n,k,alpha,lda,stride_a,beta,ldb,stride_b,ldc,stride_c,ldd,stride_d,a_type,b_type,c_type,d_type,compute_type,scaleA,scaleB,scaleC,scaleD,amaxD,activation_type,bias_vector,bias_type,rotating_buffer,hipblaslt-Gflops,hipblaslt-GB/s,us,soulution_index

N,N,0,1,64,64,1024,1.1,64,65536,0.9,1024,65536,64,4096,64,4096,f32_r,f32_r,f32_r,f32_r,f32_r,0,0,0,0,0,none,0,f32_r,512,903.229,55.8608,9.28735,49257
```

### CODE REVIEW

File: library/src/amd_detail/rocblaslt/src/UserDrivenTuningParser.cpp [hunk 1]
```diff
@@ -0,0 +1,238 @@
+#include "UserDrivenTuningParser.hpp"
+
+
+#include <fstream>
+#include <sstream>
+#include <utility>
+
+Tensile::DataType roc2TensileType(rocblaslt_compute_type type)
```
Comment (KKyang) [inline library/src/amd_detail/rocblaslt/src/UserDrivenTuningParser.cpp]: Don't copy this from tensile_host.cpp. Create a header or put this in utils.

File: library/src/amd_detail/rocblaslt/src/include/tensile_host.hpp [hunk 2]
```diff
@@ -56,6 +56,27 @@ constexpr double value_category(const T& beta)
     return beta == T(0) ? 0.0 : beta == T(1) ? 1.0 : beta == T(-1) ? -1.0 : 2.0;
 }
 
+struct TensileDataGemm
```
Comment (KKyang) [inline library/src/amd_detail/rocblaslt/src/include/tensile_host.hpp]: Exposing this is not recommended. Let it stay in tensile_host.cpp.

File: library/src/amd_detail/hipblaslt.cpp [hunk 3]
```diff
@@ -28,18 +28,46 @@
...
+        std::string file_version = firstline.substr(pos + header.length());
+        std::cerr << "hipBLASLt git version: " << git_version << std::endl;
+        std::cerr << "override file git version: " << file_version << std::endl;
+        if (file_version == git_version)
+            return true;
+    }
+
+    override.env_mode = false;
+
+    std::cerr << "The hipBLASLt git version and the override file git version are not the same." << std::endl;
```
Comment (jichangjichang) [inline library/src/amd_detail/hipblaslt.cpp]: log_error()

File: library/src/amd_detail/rocblaslt/src/include/tensile_host.hpp (hunk 2 above)
Comment (Jay0521) [reply library/src/amd_detail/rocblaslt/src/include/tensile_host.hpp]: Thanks, the above has been reimplemented in commit [b70f458](https://github.com/ROCm/hipBLASLt/pull/1289/commits/b70f4585f0fa45ac8cc544df039795ce88535ffd)

File: library/src/amd_detail/hipblaslt.cpp (hunk 3 above)
Comment (Jay0521) [reply library/src/amd_detail/hipblaslt.cpp]: Thanks, it’s done in commit [cd8e318](https://github.com/ROCm/hipBLASLt/pull/1289/commits/cd8e318ce7b88e3ac8287abac8b702e07ce1022a)

File: library/src/amd_detail/rocblaslt/src/rocblaslt_auxiliary.cpp [hunk 4]
```diff
@@ -1582,8 +1785,31 @@ rocblaslt_status
...
+                                                              workspaceBytes,
+                                                              override_result, 
+                                                              override.file_path);
+                                                              
+            log_api(__func__, "returnAlogCount", override_success ? 1 : 0);
+
+        }
+        
+        if (requestedAlgoCount - override_result.size() > 0)
+            status = getBestSolutions(handle, gemmType, gemmData, workspaceBytes, override_success ? requestedAlgoCount - 1 : requestedAlgoCount, results);
```
Comment (jichangjichang) [inline library/src/amd_detail/rocblaslt/src/rocblaslt_auxiliary.cpp]: have your run clang-format?

Comment (Jay0521) [reply library/src/amd_detail/rocblaslt/src/rocblaslt_auxiliary.cpp]: it’s done in commit [97769b6](https://github.com/ROCm/hipBLASLt/pull/1289/commits/97769b6400d447ae89cf1c640560775582ac3667)

File: clients/include/argument_model.hpp [hunk 5]
```diff
@@ -252,17 +256,25 @@ class ArgumentModel
...
+            const char*   tuningEnv  = getenv("HIPBLASLT_TUNING_FILE");
+            std::string   tuningPath = tuningEnv;
+            std::ofstream file(tuningPath, std::ios::app);
+            file << value_list << delim << archName << delim << cuNum << std::endl;
         }
-        else
+
+        str << name_list << "\n" << value_list << std::endl;
+
+        if(solution_name != "")
```
Comment (jichangjichang) [inline clients/include/argument_model.hpp:273]: why did you change the detection method?

File: clients/include/testing_matmul.hpp [hunk 6]
```diff
@@ -3254,14 +3268,22 @@ void testing_matmul_with_bias(const Arguments& arg,
 
         if(heuristicResult.size() > 1)
         {
-            int32_t     solutionIndex = -1;
-            std::string solutionName  = "";
-            std::string kernelName    = "";
+            int32_t solutionIndex = hipblaslt_ext::getIndexFromAlgo(heuristicResult[best_sol].algo);
```
Comment (jichangjichang) [inline clients/include/testing_matmul.hpp]: call getIndexFromAlgo only when it is required.

Comment (Jay0521) [reply clients/include/testing_matmul.hpp]: Thanks, it’s done.

File: clients/include/testing_matmul.hpp [hunk 7]
```diff
@@ -3257,11 +3271,22 @@ void testing_matmul_with_bias(const Arguments& arg,
...
+            const char* tuningEnv     = getenv("HIPBLASLT_TUNING_FILE");
+            if(tuningEnv)
             {
                 solutionIndex = hipblaslt_ext::getIndexFromAlgo(heuristicResult[best_sol].algo);
-                solutionName  = best_s_name;
-                kernelName    = best_k_name;
+                archName      = deviceProps.gcnArchName;
+                cuNum         = std::to_string(deviceProps.multiProcessorCount);
+            }
+
```
Comment (jichangjichang) [inline clients/include/testing_matmul.hpp:3377]: if(tuningEnv ||arg.print_kernel_info)
getindex
if(tuningEnv)
get archName and CuNum
if(arg.print_kernel_info)
get solutionname and kernel name

File: clients/include/testing_matmul.hpp [hunk 8]
```diff
@@ -3257,11 +3271,22 @@ void testing_matmul_with_bias(const Arguments& arg,
...
+            if(arg.print_kernel_info)
+            {
+                if(archName == "")
```
Comment (jichangjichang) [inline clients/include/testing_matmul.hpp]: Don't use this way for checking if we need to get solutionIdex

File: clients/include/testing_matmul.hpp (hunk 7 above)
Comment (Jay0521) [reply clients/include/testing_matmul.hpp:3377]: Thanks, it’s done in commit [2e7dcc2](https://github.com/ROCm/hipBLASLt/pull/1289/commits/2e7dcc2f6f1e1acf448fcbe05488362e1acc85aa)

File: library/src/amd_detail/rocblaslt/src/UserDrivenTuningParser.cpp [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+#include <sstream>
+#include <utility>
+
+namespace TensileLite
+{
+
+    std::multimap<ProblemOverride, int> getContractionProblemsFromFile(const std::string& path)
+    {
+
+        static std::multimap<ProblemOverride, int> m_override;
```
Comment (jichangjichang) [inline library/src/amd_detail/rocblaslt/src/UserDrivenTuningParser.cpp]: Do you have protection for m_override for multi-thread cases?

Comment (Jay0521) [reply library/src/amd_detail/rocblaslt/src/UserDrivenTuningParser.cpp]: Thanks, it’s done in commit (https://github.com/ROCm/hipBLASLt/pull/1289/commits/ff3ab5fc397d4a07745b3955e3a750bdaf866624)

**Review by jichangjichang — APPROVED** [review]

**Review by KKyang — APPROVED** [review]
